x-pack/plugin/old-lucene-versions/src/main/java/org/elasticsearch/xpack/lucene/bwc/OldLuceneVersions.java

```diff
@@ -0,0 +1,157 @@
+import org.elasticsearch.Build;
+import org.elasticsearch.common.UUIDs;
+import org.elasticsearch.common.lucene.Lucene;
+import org.elasticsearch.index.IndexModule;
+import org.elasticsearch.index.engine.Engine;
+import org.elasticsearch.index.seqno.SequenceNumbers;
+import org.elasticsearch.index.shard.IndexEventListener;
+import org.elasticsearch.index.shard.IndexShard;
+import org.elasticsearch.plugins.IndexStorePlugin;
+import org.elasticsearch.plugins.Plugin;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.HashMap;
+import java.util.Map;
+
+public class OldLuceneVersions extends Plugin implements IndexStorePlugin {
+
+    @Override
+    public void onIndexModule(IndexModule indexModule) {
+        if (Build.CURRENT.isSnapshot()) {
+            indexModule.addIndexEventListener(new IndexEventListener() {
+                @Override
+                public void afterFilesRestoredFromRepository(IndexShard indexShard) {
+                    maybeConvertToNewFormat(indexShard);
+                }
+            });
+        }
+    }
+
+    private static void maybeConvertToNewFormat(IndexShard indexShard) {
+        indexShard.store().incRef();
+        try {
+            try {
+                Version version = getLuceneVersion(indexShard.store().directory());
+                // Lucene version in [7.0.0, 8.0.0)
+                if (version != null
+                    && version.onOrAfter(Version.fromBits(7, 0, 0))
+                    && version.onOrAfter(Version.fromBits(8, 0, 0)) == false) {
+                    final OldSegmentInfos oldSegmentInfos = OldSegmentInfos.readLatestCommit(indexShard.store().directory(), 7);
+                    final SegmentInfos segmentInfos = convertLucene7x(oldSegmentInfos);
+                    // write upgraded segments file
+                    segmentInfos.commit(indexShard.store().directory());
+
+                    // validate that what we have written can be read using standard path
+                    // TODO: norelease: remove this when development completes
+                    SegmentInfos segmentInfos1 = SegmentInfos.readLatestCommit(indexShard.store().directory());
+
+                    // clean older segments file
+                    Lucene.pruneUnreferencedFiles(segmentInfos1.getSegmentsFileName(), indexShard.store().directory());
+                }
+            } catch (IOException e) {
+                throw new UncheckedIOException(e);
+            }
+        } finally {
+            indexShard.store().decRef();
+        }
+    }
+
+    private static Version getLuceneVersion(Directory directory) throws IOException {
+        final String segmentFileName = SegmentInfos.getLastCommitSegmentsFileName(directory);
+        if (segmentFileName != null) {
+            long generation = SegmentInfos.generationFromSegmentsFileName(segmentFileName);
+            try (ChecksumIndexInput input = directory.openChecksumInput(segmentFileName, IOContext.READ)) {
```

Comment:
I'm not sure why a `ChecksumIndexInput` is used here? It seems that a regular `IndexInput` can be used, or maybe there's something I miss.

Reply:
This method is going away in a follow-up.